### PR TITLE
week3 - add blobstore feature

### PR DESCRIPTION
**Pull request by joeythegod**
week3 - add blobstore feature

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 1]
```diff
@@ -97,6 +112,43 @@ private String getParameter(HttpServletRequest request, String name, String defa
     }
     return value;
   }
+
+  /** Returns a URL that points to the uploaded file, or null if the user didn't upload a file. */
```
Comment (0charleschen0) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]:
```suggestion
  /** Returns a URL that points to the uploaded file, or {@code null} if the user didn't upload a file. */
```

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 2]
```diff
@@ -97,6 +112,43 @@ private String getParameter(HttpServletRequest request, String name, String defa
...
+  private String getUploadedFileUrl(HttpServletRequest request, String formInputElementName) {
```
Comment (0charleschen0) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:120]:
```suggestion
  @Nullable
  private String getUploadedFileUrl(HttpServletRequest request, String formInputElementName) {
```
Add @Nullable to make sure callers have null checks.

File: portfolio/src/main/webapp/script.js [hunk 3]
```diff
@@ -47,14 +47,37 @@ function createListElement(text) {
...
+  fetch('/blobstore-upload-url')
+      .then((response) => {
+        return response.text();
+      })
+      .then((imageUploadUrl) => {
+        const messageForm = document.getElementById('my-form');
+        messageForm.action = imageUploadUrl;
+        messageForm.classList.remove('hidden');
+      });
 }
```
Comment (0charleschen0) [inline portfolio/src/main/webapp/script.js]: Super nit: Can add an empty line below so we don't get distracted by the mark.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 4]
```diff
@@ -74,11 +87,13 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
     String name = getParameter(request, "text-input-name", "");
     String content = getParameter(request, "text-input-content", "");
+    String imageUrl = getUploadedFileUrl(request, "imageUrl");
     long timestamp = System.currentTimeMillis();
 
     Entity commentEntity = new Entity("comment");
     commentEntity.setProperty("name", name);
     commentEntity.setProperty("content", content);
+    commentEntity.setProperty("imageUrl", imageUrl);
```
Comment (0charleschen0) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:98]: Could you please verify that we don't get NullPointerException if imageUrl is null?

**Review by 0charleschen0 — COMMENTED** [review]
Good job! Perfect code style! Just some comments about null check handling.

File: portfolio/src/main/webapp/script.js (hunk 3 above)
Comment (joeythegod) [reply portfolio/src/main/webapp/script.js]: OK!

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 4 above)
Comment (joeythegod) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:98]: Verified!

**Review by 0charleschen0 — APPROVED** [review]

**Pull request merged**